examples/generative/dreambooth.py

```diff
@@ -0,0 +1,672 @@
+
+instance_image_paths, class_image_paths[:5]
+
+"""
+Then we load the images from the paths. 
+"""
+
+from PIL import Image
+import numpy as np
+
+
+def load_images(image_paths):
+    images = []
+    for path in image_paths:
+        image = Image.open(path)
+        images.append(np.array(image))
+    return images
+
+
+"""
+And then we make use a utility function to plot the loaded images. 
+"""
+
+
+import matplotlib.pyplot as plt
+
+
+def plot_images(images, title=None):
+    plt.figure(figsize=(20, 20))
+    for i in range(len(images)):
+        ax = plt.subplot(1, len(images), i + 1)
+        if title is not None:
+            plt.title(title)
+        plt.imshow(images[i])
+        plt.axis("off")
+
+
+"""
+**Instance images**:
+"""
+
+plot_images(load_images(instance_image_paths[:5]))
+
+"""
+**Class images**:
+"""
+
+plot_images(load_images(class_image_paths[:5]))
+
+"""
+## Prepare datasets
+
+Dataset preparation includes two stages: (1): preparing the captions, (2) processing the
+images.
+"""
+
+"""
+### Prepare the captions
+"""
+
+# Since we're using prior preservation, we need to match the number
+# of instance images we're using. We just repeat the instance image paths
+# to do so.
+new_instance_image_paths = []
+for index in range(len(class_image_paths)):
+    instance_image = instance_image_paths[index % len(instance_image_paths)]
+    new_instance_image_paths.append(instance_image)
+
+# We just repeat the prompts / captions per images.
+unique_id = "sks"
+class_label = "dog"
+
+instance_prompt = f"a photo of {unique_id} {class_label}"
+instance_prompts = [instance_prompt] * len(new_instance_image_paths)
+
+class_prompt = f"a photo of {class_label}"
+class_prompts = [class_prompt] * len(class_image_paths)
+
+"""
+Next, we embed the prompts to save some compute.  
+"""
+
+import itertools
+
+# The padding token and maximum prompt length are specific to the text encoder.
+# If you're using a different text encoder be sure to change them accordingly.
+padding_token = 49407
+max_prompt_length = 77
+
+# Load the tokenizer.
+tokenizer = SimpleTokenizer()
+
+# Method to tokenize and pad the tokens.
+def process_text(caption):
+    tokens = tokenizer.encode(caption)
+    tokens = tokens + [padding_token] * (max_prompt_length - len(tokens))
+    return np.array(tokens)
+
+
+# Collate the tokenized captions into an array.
+tokenized_texts = np.empty(
+    (len(instance_prompts) + len(class_prompts), max_prompt_length)
+)
+
+for i, caption in enumerate(itertools.chain(instance_prompts, class_prompts)):
+    tokenized_texts[i] = process_text(caption)
+
+
+# We also pre-compute the text embeddings to save some memory during training.
+POS_IDS = tf.convert_to_tensor([list(range(max_prompt_length))], dtype=tf.int32)
+text_encoder = TextEncoder(max_prompt_length)
+
+gpus = tf.config.list_logical_devices("GPU")
+
+# Ensure the computation takes place on a GPU.
+with tf.device(gpus[0].name):
```

Comment:
This should happen automatically, no?

Reply:
It should. I just wanted to show how to do perform device placement during data loading as well.

Reply:
Please add a comment, otherwise a reader may assume that adding this line is required.